### ISSUE TITLE
NPE in codegen when running Android kotlin tests

### ISSUE THREAD

**Issue by edenman** [google/dagger#1221]
I've been trying to create a sample project to repro, but so far i've been unsuccessful.

```
ext.dagger_version = '2.16'
  ext.kotlin_version = '1.2.50'
  ext.support_lib_version = '1.0.0-alpha3'
    classpath 'com.android.tools.build:gradle:3.3.0-alpha03'
  compileSdkVersion 28
    minSdkVersion 21
    targetSdkVersion 28
...
  sourceSets {
    main.java.srcDirs += 'src/main/kotlin'
    debug.java.srcDirs += 'src/debug/kotlin'
    test.java.srcDirs += 'src/test/kotlin'
  }
junit 4.12
robolectric 4.0-alpha-3-SNAPSHOT (to get support for 28)
```
I can't figure out from MembersInjectionMethods:74 what's actually going wrong.  It's possible the non-existent classpath locations are the issue, or maybe that's just a red herring? 
```
> Task :kaptGenerateStubsDebugUnitTestKotlin UP-TO-DATE
w: Classpath entry points to a non-existent location: /Users/edenman/Development/recharge/src/android/consumer/recharge-consumer-android/src/test/java
w: Classpath entry points to a non-existent location: /Users/edenman/Development/recharge/src/android/consumer/recharge-consumer-android/src/testDebug/kotlin
w: Classpath entry points to a non-existent location: /Users/edenman/Development/recharge/src/android/consumer/recharge-consumer-android/src/testDebug/java
e: [kapt] An exception occurred: java.lang.NullPointerException
	at dagger.internal.codegen.MembersInjectionMethods.membersInjectionMethod(MembersInjectionMethods.java:74)
	at dagger.internal.codegen.Util.reentrantComputeIfAbsent(Util.java:94)
	at dagger.internal.codegen.MembersInjectionMethods.getOrCreate(MembersInjectionMethods.java:68)
	at dagger.internal.codegen.MembersInjectionBindingExpression.membersInjectionInvocation(MembersInjectionBindingExpression.java:68)
	at dagger.internal.codegen.MembersInjectionBindingExpression.getComponentMethodImplementation(MembersInjectionBindingExpression.java:62)
	at dagger.internal.codegen.ComponentBindingExpressions.getComponentMethod(ComponentBindingExpressions.java:247)
	at dagger.internal.codegen.ComponentWriter.addInterfaceMethods(ComponentWriter.java:195)
	at dagger.internal.codegen.ComponentWriter.write(ComponentWriter.java:169)
	at dagger.internal.codegen.ComponentWriter.writeSubcomponent(ComponentWriter.java:118)
	at dagger.internal.codegen.ComponentWriter.addSubcomponents(ComponentWriter.java:206)
	at dagger.internal.codegen.ComponentWriter.write(ComponentWriter.java:170)
	at dagger.internal.codegen.ComponentWriter.writeComponent(ComponentWriter.java:92)
	at dagger.internal.codegen.ComponentGenerator.write(ComponentGenerator.java:73)
	at dagger.internal.codegen.ComponentGenerator.write(ComponentGenerator.java:34)
	at dagger.internal.codegen.SourceFileGenerator.generate(SourceFileGenerator.java:69)
	at dagger.internal.codegen.SourceFileGenerator.generate(SourceFileGenerator.java:60)
	at dagger.internal.codegen.ComponentProcessingStep.generateComponent(ComponentProcessingStep.java:168)
	at dagger.internal.codegen.ComponentProcessingStep.process(ComponentProcessingStep.java:147)
	at dagger.internal.codegen.ComponentProcessingStep.process(ComponentProcessingStep.java:46)
	at dagger.shaded.auto.common.BasicAnnotationProcessor.process(BasicAnnotationProcessor.java:330)
	at dagger.shaded.auto.common.BasicAnnotationProcessor.process(BasicAnnotationProcessor.java:181)
	at org.jetbrains.kotlin.kapt3.base.ProcessorWrapper.process(annotationProcessing.kt:99)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1068)
	at org.jetbrains.kotlin.kapt3.base.AnnotationProcessingKt.doAnnotationProcessing(annotationProcessing.kt:55)
	at org.jetbrains.kotlin.kapt3.base.AnnotationProcessingKt.doAnnotationProcessing$default(annotationProcessing.kt:27)
	at org.jetbrains.kotlin.kapt3.AbstractKapt3Extension.runAnnotationProcessing(Kapt3Extension.kt:216)
	at org.jetbrains.kotlin.kapt3.AbstractKapt3Extension.analysisCompleted(Kapt3Extension.kt:164)
	at org.jetbrains.kotlin.kapt3.ClasspathBasedKapt3Extension.analysisCompleted(Kapt3Extension.kt:93)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM$analyzeFilesWithJavaIntegration$2.invoke(TopDownAnalyzerFacadeForJVM.kt:98)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration(TopDownAnalyzerFacadeForJVM.kt:108)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration$default(TopDownAnalyzerFacadeForJVM.kt:85)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:370)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:61)
	at org.jetbrains.kotlin.cli.common.messages.AnalyzerWithCompilerReport.analyzeAndReport(AnalyzerWithCompilerReport.kt:101)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.analyze(KotlinToJVMBytecodeCompiler.kt:361)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:126)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:154)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:51)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:95)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:50)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:88)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$compile$1$1$2.invoke(CompileServiceImpl.kt:405)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$compile$1$1$2.invoke(CompileServiceImpl.kt:98)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$doCompile$$inlined$ifAlive$lambda$2.invoke(CompileServiceImpl.kt:927)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$doCompile$$inlined$ifAlive$lambda$2.invoke(CompileServiceImpl.kt:98)
	at org.jetbrains.kotlin.daemon.common.DummyProfiler.withMeasure(PerfUtils.kt:137)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.checkedCompile(CompileServiceImpl.kt:957)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.doCompile(CompileServiceImpl.kt:926)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:404)
	at sun.reflect.GeneratedMethodAccessor98.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:324)
	at sun.rmi.transport.Transport$1.run(Transport.java:200)
	at sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:568)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:826)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:683)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:682)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by ronshapiro** [comment]
Can you update to include the version of Dagger?

**Comment by edenman** [comment]
markdown derp, fixed

**Comment by ronshapiro** [comment]
And did you recently add a members injection method on your component?

**Comment by edenman** [comment]
I haven't changed anything about my dagger setup lately.  I have updated quite a few things (androidx, kotlin, AGP, etc) but I don't know which of those is to blame bc my CI has been broken for a while and I don't regularly run tests myself :/

**Comment by edenman** [comment]
I tried removing all the new members injection methods on my component.  No luck.  Added some more info to the crash: https://github.com/google/dagger/pull/1228

**Comment by edenman** [comment]
Just realized I could install a local snapshot of my fork to see what's going on.  Looks like it's blowing up on injecting an inner class that I added injections for _ages_ ago.  Maybe something about the format of inner classes changed in a kotlin upgrade?

**Comment by edenman** [comment]
Something is causing the lookup to fail.  I added a bunch more printlns:

```
- Key.hashCode: 549960818 vs -783806129
- Key.equals: true
- type: co.recharge.consumer.MapController.AuthenticatedMapController vs co.recharge.consumer.MapController.AuthenticatedMapController TypeMirror.equals? false 
- qualifier: Optional.empty vs Optional.empty qualifier.equals? true 
- multibindingContributionIdentifier: Optional.empty vs Optional.empty multibindingContributionIdentifier.equals? true 
- wrappedType: dagger.shaded.auto.common.MoreTypes$TypeEquivalence@3fc54cc8.wrap(co.recharge.consumer.MapController.AuthenticatedMapController) vs dagger.shaded.auto.common.MoreTypes$TypeEquivalence@3fc54cc8.wrap(co.recharge.consumer.MapController.AuthenticatedMapController) Equivalence.Wrapper.equals? true
- wrappedType.hashCode: 1096245132 vs 1096245132
```
Here's the printlns so you can see exactly what I'm printing out
```
        allBindingsStr.append("\n Key.hashCode: " + daKey.hashCode() + " vs " + key.hashCode());
        allBindingsStr.append("\n Key.equals: " + daKey.equals(key));
        allBindingsStr.append("\n type: " + daKey.type() + " vs " + key.type() + " TypeMirror.equals? " + daKey.type().equals(key.type()) + " ");
        allBindingsStr.append("\n qualifier: " + daKey.qualifier() + " vs " + key.qualifier() + " qualifier.equals? " + daKey.qualifier().equals(key.qualifier()) + " ");
        allBindingsStr.append("\n multibindingContributionIdentifier: " + daKey.multibindingContributionIdentifier() + " vs " + key.multibindingContributionIdentifier() + " multibindingContributionIdentifier.equals? " + daKey.multibindingContributionIdentifier().equals(key.multibindingContributionIdentifier()) + " ");
        Equivalence.Wrapper<TypeMirror> wrappedDaKey = MoreTypes.equivalence().wrap(daKey.type());
        Equivalence.Wrapper<TypeMirror> wrappedKey = MoreTypes.equivalence().wrap(key.type());
        allBindingsStr.append("\n wrappedType: " + wrappedDaKey + " vs " + wrappedKey + " Equivalence.Wrapper.equals? " + wrappedDaKey.equals(wrappedKey));
        allBindingsStr.append("\n wrappedType.hashCode: " + wrappedDaKey.hashCode() + " vs " + wrappedKey.hashCode());
```

**Comment by edenman** [comment]
I got super annoyed by this still being an issue so I extracted the class to be a non-inner class...and another inner class I'm injecting started blowing up.  Finally got a repro project up and running:
https://github.com/edenman/DaggerKotlinTestFail

**Comment by edenman** [comment]
Was able to hook up the debugger using `./gradlew --no-daemon -Dorg.gradle.debug=true -Dkotlin.compiler.execution.strategy="in-process" :app:test` (thanks ASG!)

Looks like the issue is that the `com.sun.tools.javac.code.Type$ClassType` is subtly different: `allparams_field` is null vs empty list, same with `interfaces_field`.  And `supertype_field` is `java.lang.Object` in the candidate Key, and null in the `membersInjectionBindings` map.  This results in the hashCode being different, so the map lookup fails.

![screen shot 2018-07-31 at 11 50 23 am](https://user-images.githubusercontent.com/1063557/43480599-9e009a62-94b8-11e8-8f6e-cbe792fa09e6.png)
![screen shot 2018-07-31 at 11 50 33 am](https://user-images.githubusercontent.com/1063557/43480600-9e1ae8a4-94b8-11e8-8d2a-ba54fa76034b.png)

Anybody have ideas about what could be causing the mismatch?

**Comment by edenman** [comment]
Ok I thought it was bc `Type$ClassType` has public mutable fields AND has lazy accessor methods for some of its fields (`allparams()`).  But `MoreTypes.HashVisitor` only seems to take into account:

1) `TypeMirror.asElement()`
2) `TypeMirror.getEnclosingType()`
3) `TypeMirror.getTypeArguments()`

`asElement()` returns a `Symbol$ClassSymbol` which afaict doesn't actually implement `hashCode()`...so we're relying on the object id?  This seems like a bad thing but I have no idea why this would only be an issue for inner classes.

**Comment by ronshapiro** [comment]
@edenman sorry I haven't been chiming in here much, but I'm pretty sure that javac's API/implementation guarantees that there will never be (unless we're across rounds) different instances of the same symbol.

**Comment by edenman** [comment]
@ronshapiro yup, just looked again at the debugger screenshots and both `tsym` objects have the same id, so that theory is busted.

**Comment by edenman** [comment]
So I'm starting to think my initial theory about `Type$ClassType` being mutable is maybe the root of the issue, but I can't tell why yet.  All of the component hashCodes are the same, but the memoized hashCode is different, which seems to be a sign that one of the underlying objects was mutated after its hashCode was initially computed:

![screen shot 2018-07-31 at 2 41 06 pm](https://user-images.githubusercontent.com/1063557/43488840-cb4f9678-94cf-11e8-8b7b-9ba78269478f.png)

**Comment by edenman** [comment]
Ok so that theory was also wrong: there's apparently a difference in how the inner class symbol is being passed in:
![screen shot 2018-07-31 at 3 30 37 pm](https://user-images.githubusercontent.com/1063557/43490770-cf0dc242-94d6-11e8-99f9-780fb20df74d.png)
![screen shot 2018-07-31 at 3 31 07 pm](https://user-images.githubusercontent.com/1063557/43490771-cf2503d0-94d6-11e8-9c9c-e1d311889b38.png)

**Comment by ronshapiro** [comment]
What happened to your comment about inner class symbols being different? I got the email but I don't see the github log.

If that's the case, it seems like a kapt bug that may be changing how `Taco` is being passed to javac?

**Comment by edenman** [comment]
Yeah, sorry, I realized it wasn't 100% correct after posting.  Here we go:

![screen shot 2018-07-31 at 3 30 37 pm](https://user-images.githubusercontent.com/1063557/43491805-e615ace4-94da-11e8-8007-6b58c3751e39.png)
![screen shot 2018-07-31 at 3 31 07 pm](https://user-images.githubusercontent.com/1063557/43491806-e62fa3e2-94da-11e8-8aae-64b645704fee.png)

The reason this is the case is that one of them still has a `completer` set on the `tsym`.  If you eagerly call the completer and then construct the Key, this problem goes away.  But you're probably right that the fix needs to go in kapt?

**Comment by edenman** [comment]
I still don't entirely understand why they're different `ClassSymbol` instances but maybe the cache is looking them up by the fullname or something?

**Comment by edenman** [comment]
Ok so the different instances thing was because the codegen was getting run twice (`debugKotlin`, `debugUnitTestKotlin`).  Switching back to println debugging:

```
> Task :app:kaptDebugKotlin
forInjectConstructorWithResolvedType GOT ClassType 515476882
forInjectConstructorWithResolvedType GOT fullname org.coffeetrain.daggerkotlintestfail.MainActivity.Taco
forInjectConstructorWithResolvedType GOT flat_name org.coffeetrain.daggerkotlintestfail.MainActivity$Taco
forInjectConstructorWithResolvedType GOT members_field Scope[Taco(), setFoo(org.coffeetrain.daggerkotlintestfail.Foo), getFoo(), foo]
forInjectConstructorWithResolvedType GOT hasCompleter false
forMembersInjectedType GOT ClassType 515476882
forMembersInjectedType GOT fullname org.coffeetrain.daggerkotlintestfail.MainActivity.Taco
forMembersInjectedType GOT flat_name org.coffeetrain.daggerkotlintestfail.MainActivity$Taco
forMembersInjectedType GOT members_field Scope[Taco(), setFoo(org.coffeetrain.daggerkotlintestfail.Foo), getFoo(), foo]
forMembersInjectedType GOT hasCompleter false
forMembersInjectedType GOT ClassType 515476882
forMembersInjectedType GOT fullname org.coffeetrain.daggerkotlintestfail.MainActivity.Taco
forMembersInjectedType GOT flat_name org.coffeetrain.daggerkotlintestfail.MainActivity$Taco
forMembersInjectedType GOT members_field Scope[Taco(), setFoo(org.coffeetrain.daggerkotlintestfail.Foo), getFoo(), foo]
forMembersInjectedType GOT hasCompleter false
w: Classpath entry points to a non-existent location: /Users/edenman/Development/DaggerKotlinTestFail/app/src/debug/kotlin
w: Classpath entry points to a non-existent location: /Users/edenman/Development/DaggerKotlinTestFail/app/build/generated/aidl_source_output_dir/debug/compileDebugAidl/out
w: Classpath entry points to a non-existent location: /Users/edenman/Development/DaggerKotlinTestFail/app/src/main/java
w: Classpath entry points to a non-existent location: /Users/edenman/Development/DaggerKotlinTestFail/app/src/debug/java

> Task :app:kaptDebugUnitTestKotlin
forMembersInjectedType GOT ClassType 660086630
forMembersInjectedType GOT fullname org.coffeetrain.daggerkotlintestfail.MainActivity$Taco
forMembersInjectedType GOT flat_name org.coffeetrain.daggerkotlintestfail.MainActivity$Taco
forMembersInjectedType GOT members_field null
forMembersInjectedType GOT hasCompleter true
forInjectConstructorWithResolvedType GOT ClassType 660086630
forInjectConstructorWithResolvedType GOT fullname org.coffeetrain.daggerkotlintestfail.MainActivity.Taco
forInjectConstructorWithResolvedType GOT flat_name org.coffeetrain.daggerkotlintestfail.MainActivity$Taco
forInjectConstructorWithResolvedType GOT members_field Scope[Taco(), setFoo(org.coffeetrain.daggerkotlintestfail.Foo), getFoo(), foo]
forInjectConstructorWithResolvedType GOT hasCompleter false
forMembersInjectedType GOT ClassType 660086630
forMembersInjectedType GOT fullname org.coffeetrain.daggerkotlintestfail.MainActivity.Taco
forMembersInjectedType GOT flat_name org.coffeetrain.daggerkotlintestfail.MainActivity$Taco
forMembersInjectedType GOT members_field Scope[Taco(), setFoo(org.coffeetrain.daggerkotlintestfail.Foo), getFoo(), foo]
forMembersInjectedType GOT hasCompleter false
```

**Comment by edenman** [comment]
Ok one more comment then I'm done for the day: should I file this as a kapt bug?  I can't see where the issue would be on their side, but I'm not super familiar with the contract between kapt and annotation processors.

What's interesting to me is that the `completer` has already been called in `app:kaptDebugKotlin`, and the call sequence is `forInjectConstructorWithResolvedType`,`forMembersInjectedType`,`forMembersInjectedType`.

Then in `app:kaptDebugUnitTestKotlin`, the `completer` hasn't been called, and the call sequence is `forMembersInjectedType`,`forInjectConstructorWithResolvedType`,`forMembersInjectedType`.

**Comment by edenman** [comment]
Filed a kapt bug, we'll see if that goes anywhere: https://youtrack.jetbrains.com/issue/KT-25923

**Comment by edenman** [comment]
On a tip from @ronshapiro, I confirmed that converting `MyComponent` to java fixes the issue: https://github.com/edenman/DaggerKotlinTestFail/compare/edenman/java-component

**Comment by edenman** [comment]
If anybody else hits this issue: I've made a fork with the previously-mentioned workaround and I'm depending on that in my apps until the kapt bug gets fixed:

add new repo: https://github.com/rechargelabs/mvn-peripheral/raw/3881575e93ecf1f3c9142fcc834f45e0c6e85da9/
change dagger version to `2.17-edenman-kapt-bug-workaround`

**Comment by ronshapiro** [comment]
@edenman do you think it's safe to close this as not being a Java issue?

**Comment by edenman** [comment]
I'd probably err on the side of leaving it open so it's easier to find if anybody else runs into it, but up to you.

**Comment by yanex** [comment]
@edenman Unfortunately, I couldn't reproduce the bug with your example (https://github.com/edenman/DaggerKotlinTestFail). Can you please check if it's actual, and if not, provide any other?

**Comment by fengxiaocan** [comment]
Hi，How did you finally solve this problem? I had this problem too, but I couldn't find anyone to describe it。

**Comment by edenman** [comment]
@fengxiaocan see workaround in https://github.com/google/dagger/issues/1221#issuecomment-413258421

@yanex do you need any more information?  were you able to repro in my sample repo?

**Comment by edenman** [comment]
This seems to be fixed as of Dagger 2.21
